File: test_unigram.py
#! /usr/bin/env python
# coding:utf-8

from train_unigram import train
from collections import defaultdict
import math


def test(
    modelfile: str,
    filename: str,
    lambda1: float=0.95,
    n_unk: int=1e6
    ):
    # load model
    model = {}
    with open(modelfile) as f:
        for key, val in (line.split("\t") for line in f):
            model[key] = float(val)

    entropy = 0
    total = 0
    found = 0

    with open(filename) as f:
        for line in f:
            for word in line.split() + ["</s>"]:
                total += 1
                if word in model:
                    ent = lambda1 * model[word]
                    found += 1
                else:
                    ent = (1 - lambda1) * (1 / n_unk)
                entropy += -math.log2(ent)
    entropy = entropy / total

    print("entropy = {:.6f}".format(entropy))
    print("coverage = {:.6f}".format(found / total))



if __name__ == '__main__':
    import sys
    modelfile, filename = sys.argv[1:3]
    test(modelfile, filename)

File: test_unigram.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <fstream>
#include <string>
#include <stack>
#include <vector>
#include <map>
#include <boost/algorithm/string.hpp>
#include <cmath>

#include "util.h"

using namespace std;


void test(
        const char* modelfile,
        const char* filename,
        const double lambda1 = 0.95,
        const long n_unk = 1e6
) {
    // load model
    map<string, double> model = load_unimodel(modelfile);

    ifstream datafs(filename);
    string line;
    long total = 0;
    long found = 0;
    double entropy = 0;
    while (getline(datafs, line)) {
        vector<string> words;
        boost::algorithm::split(words, line, boost::is_space());
        words.push_back("</s>");
        double prob = 0;

        prob = uniprob(words[0], model, lambda1, n_unk);
        for (auto word: words) {
            if (model.count(word)) {
                found++;
            }
            total++;
            prob = uniprob(word, model, lambda1, n_unk);
            entropy -= log2(prob);
        }
    }
    entropy = entropy / total;
    double coverage = (double)found / total;

    printf("entropy = %f\ncoverage = %f\n", entropy, coverage);
}


int main(int argc, char const* argv[])
{
    const char *modelfile = argv[1];
    const char *filename = argv[2];
    test(modelfile, filename);
    return 0;
}

